### ISSUE TITLE
Partial<T>[K] is not assignable to T[K] | undefined

### ISSUE THREAD

**Issue by hearnden** [microsoft/TypeScript#47523]
# Bug Report

...ok, I lied a little:
 - it _is_ assignable if `T` is a type variable / generic; but
 - it _is not_ assignable if `T` is concrete.

### 🔎 Search Terms

Searching for:
 - "Partial"
 - "Partial undefined"

I found these potentially related issues:
 - https://github.com/microsoft/TypeScript/issues/31675
 - https://github.com/microsoft/TypeScript/issues/45257#issuecomment-890460184

#45257 looks very similar.  However, I think this report is a simpler case, since seems more directly related to the definition of `Partial<>` / optional properties (`?`), as something that adds `| undefined` to the type of a property access expression.  It also looks to me like a fix for this report would fix #45257 naturally.

#31675 also looks related, but from the non-nullable perspective rather than the nullable one.  The non-nullable inference rule also appears to be missing from _both_ generic inference and concrete inference (e.g., `NonNullable<Partial<T>[K]>` is not assignable for `T[K]` for both generic `T` and concrete `T`), so at least there isn't an inconsistency there.

### 🕗 Version & Regression Information

Worked in 3.3.3
Broken in: 3.5.1

This changed between versions 3.3.3 and 3.5.1.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4#code/C4TwDgpgBAYg9nKBeKBvAUFLUAeAuKAOwFcBbAIwgCcAaTbEAgZ2CoEtCBzOgXwG506AGbFCAY2Bs4hKJwjAAagEMANsQgBxCIWpsxAHgAqNKAGkoEHMG0ATJlADWEEHCFRDAPgAUcAgAUlKklVIw8TBwJTAEoCQwBtUwBdKAAfKFEbCCEOCBsMbCgqeWIqGTg4h0SBHkERcUlpWXllNQgAYWkxIut9c0trQjtHZ1dYBG9fKACgthD4ODDHSJix8qTU9MGsnJs0eixukrKKqvQaoA)

### 💻 Code

```ts
type Foo = {
    x: number,
    y: string,
};

/** Reads a key of any Partial<T> as an optional. */
function getValueGeneric<T, K extends keyof T>(o: Partial<T>, k: K): T[K] | undefined {
    return o[k];  // Ok: Type 'Partial<T>[K]' is assignable to type 'T[K] | undefined'
}

/** Reads a key of a Partial<Foo> as an optional. */
function getValueConcrete<K extends keyof Foo>(o: Partial<Foo>, k: K): Foo[K] | undefined {
    return o[k];  // Error: Type 'Partial<Foo>[K]' is not assignable to type 'Foo[K] | undefined'
}
```

### 🙁 Actual behavior

 - `Partial<T>[K]` is assignable to `T[K] | undefined` when `T` is a generic
 - `Partial<T>[K]` is not assignable to `T[K] | undefined` when `T` is concrete

### 🙂 Expected behavior

 - Assignability of `Partial<T>[K]` to `T[K] | undefined` should be consistent for both generic `T` and concrete `T`
 - Both should be allowed

**Comment by fatcerberus** [comment]
Possibly this is due to #30769?

**Comment by RyanCavanaugh** [comment]
Indexed access on a mapped type stays deferred, by design (since this improves the specificness of many constructs). The only way to know that `{ [K in keyof T]: T[K] | undefined }[K]` is assignable to `T[K] | undefined` is to know that *at a higher level* this mapped type doesn't do something else like `{ [K in keyof T]: T[K] | string | undefined }[K]`. We can do some reaching backwards but it's not a generalizable principle, and pushes the inconsistency further down the line rather than remove it.

An equivalent[ish] definition that works well is
```ts
type Foo = { x: string, y: number };
function getValueConcrete<P extends Partial<Foo>, K extends (keyof Foo) & (keyof P)>(o: P, k: K): P[K] | undefined {
    return o[k];
}
```
this has the advantage that nonsense reads like
```ts
getValueConcrete({x: "" }, "y");
```
fail

**Comment by yume-chan** [comment]
Does that mean the generic example doesn't report an error is because TypeScript didn't check it (deferred)?

---

The error message is cryptic:

```text
Type 'Partial<Foo>[K]' is not assignable to type 'Foo[K] | undefined'.
  Type 'string | number | undefined' is not assignable to type 'Foo[K] | undefined'.
    Type 'string' is not assignable to type 'Foo[K] | undefined'.
      Type 'Partial<Foo>[K]' is not assignable to type 'Foo[K]'.
        Type 'number | undefined' is not assignable to type 'number'.
          Type 'undefined' is not assignable to type 'number'.
```

---

This usage looks very common to me, for example I may write:

```ts
type FooOptions = {
    x: number,
    y: string,
};

class Foo {
    constructor(private options: Partial<FooOptions>) { }

    getOption<K extends keyof FooOptions>(key: K): FooOptions[K] | undefined {
        return this.options[key]; // Error
    }
}
```

---
